File: hgle_program3.jl
# program 3 - Dataframes!!'

# include packages
using CSV, DataFrames

function main()
    println("This is going to be an aswesome use of DataFrames... REALY!!!")

    person_file_name = "/oscar/data/shared/ursa/synthetic_ri/demo_omop/person.csv"

    patient_df = DataFrame(CSV.File(person_file_name))

    # println(describe(patient_df))

    # print out the first 7 rows (all columns)
    #println("print out the first 7 rows")
    #println(patient_df[1:7,:])
    #println(first(patient_df, 7))

    #=
    unique_race_concept_ids = unique(patient_df.race_concept_id)
    unique_ethnicity_concept_ids = unique(patient_df.ethnicity_concept_id)

    race_set = Set(unique_race_concept_ids)
    ethnicity_set = Set(unique_ethnicity_concept_ids)

    race_ethnicity_set = union(race_set, ethnicity_set)

    println("race set: $race_set")
    println("ethnicity set: $ethnicity_set")
    println("union set: $race_ethnicity_set")
    =#
    
    # retrieves all the unique calues of brith month into a set
    unique_months = unique(patient_df.month_of_birth)
    println(unique_months)

    #=
    # report count of patients per month
    for month in unique_months

        month_df = patient_df[patient_df.month_of_birth .== month, :]
        println("number of patients in $month: $(nrow(month_df))")
        
    end
    =#

        month_dict = Dict(1 => "Jan", 
                          2 => "Feb", 
                          3 => "Mar", 
                          4 => "Apr", 
                          5 => "May", 
                          6 => "Jun", 
                          7 => "Jul", 
                          8 => "Aug", 
                          9 => "Sep", 
                          10 => "Oct", 
                          11 => "Nov", 
                          12 => "Dec")

    for month in unique_months

        month_df = patient_df[patient_df.month_of_birth .== month, :]
        println("number of patients in $(month_dict[month]): $(nrow(month_df))")

    end

    month_array = 1:12

    for month in month_array

        month_df = patient_df[patient_df.month_of_birth .== month, :]
        println("number of patients in $(month_dict[month]): $(nrow(month_df))")

    end

    # report top 5 months of birth
    top_months = combine(groupby(patient_df, :month_of_birth), nrow => :count)
    sorted_months = sort(top_months, :count, rev=true)
    
    for row in eachrow(first(sorted_months, 5))
        month = row.month_of_birth
        count = row.count

        println("For $month, there are $count patients")
    end

end

main()

File: hgle_program3.py
# program3.py
# Dataframe using Pandas

import pandas as pd 

def main():
    print("Hi! this is my cool pandas program!")

    person_file_name = "/oscar/data/shared/ursa/synthetic_ri/demo_omop/person.csv"
    patient_df       = pd.read_csv(person_file_name)

    print(patient_df.describe())

    # select first 7 rows (all columns)
    print("first seven rows (all columns)")
    print(patient_df.iloc[7])
    #print(patient_df.iloc[:7])

    # retrieves race/eth concept_ids
    unique_race_concept_ids      = patient_df["race_concept_id"].unique()
    unique_ethnicity_concept_ids = patient_df["ethnicity_concept_id"].unique()

    # convert the unique values to sets
    race_set      = set(unique_race_concept_ids)
    ethnicity_set = set(unique_ethnicity_concept_ids)

    # take the union of the race and ethnicity sets
    race_ethnicity_set = race_set | ethnicity_set

    # print out set and their union
    print(f"race set:       {race_set}")
    print(f"ethnicity set:  {ethnicity_set}")
    print(f"union set:      {race_ethnicity_set}")

    # retrieve all unique values in birth month and load into a set
    unique_months = patient_df["month_of_birth"].unique()
    month_set     = set(unique_months)
    print(f"unique birth months: {month_set}")

    # report count patients born per month
    for month in month_set:
        month_df = patient_df[patient_df["month_of_birth"] == month]
        print(f"num patients in {month}: {len(month_df)}")

    # create dictionary to map month numbers to month names
    month_dict = {
        1: "Jan", 
        2: "Feb", 
        3: "Mar", 
        4: "Apr", 
        5: "May", 
        6: "Jun", 
        7: "Jul", 
        8: "Aug", 
        9: "Sep", 
        10: "Oct", 
        11: "Nov", 
        12: "Dec" 
    }

    # create report range
    month_array = range(6,9)

    # report count of patients per selected range month
    for month in month_array:
        month_df = patient_df[patient_df["month_of_birth"] == month]
        print(f"[SUM] num patients in {month_dict[month]}: {len(month_df)}")

    # find the top 3 months of birth
    top_months = patient_df.groupby("month_of_birth").size().reset_index(name="count")
    sorted_months = top_months.sort_values("count", ascending=False)

    # print out the top 7
    for _, row in sorted_months.head(7).iterrows():
        month = row["month_of_birth"]
        count = row["count"]
        print(f"For {month}: there are {count} patients")

main()